Clarification of CNF Conformance Program, adding CNF Working Group info (#450)
cncf/cnf-conformance: cnf-wg/charter.md

# Cloud Native Network Functions Working Group Charter

## Introduction

> The goal of the Cloud Native Network Functions Working Group (CNF WG)  is to aid companies such as communications service providers and large scale enterprises, running internal telecommunications-like infrastructure, to better understand what cloud native means for telecommunications workloads and help build consensus around industry adoption of cloud native technologies (per TUG whitepaper).

Reviewer: ```suggestion
The goal of the Cloud Native Network Functions Working Group (CNF WG)  is to aid companies such as telecom vendors, communications service providers and large scale enterprises, running internal telecommunications-like infrastructure, to better understand what cloud native means for telecommunications workloads and help build consensus around industry adoption of cloud native technologies (per CNCF TUG).
```

I (:)) suprisingly think that this is most important for the ones who are building the CNF-s, the telecom vendors. 

I think we should have a TUG whitepaper first before we refer to it. In my opinion CNCF TUG never finished agreeing in one WP. We have different bits and pieces written in different pieces by different members of CNCF TUG, but I would not call any of them the TUG WP.

Author: That paragraph is referring to what the TUG white-paper did cover which was the intent.  

The TUG can create new white-papers which may have a broader scope or whatever it deems useful.

This WG will be defining what it means for a telco/network application to adhere to cloud native best practices and principles as well as the process for certification.

The CNF WG operates under the aegis of CNCF. The charter of the working group is to define the process around certifying the cloud nativeness of telco applications, aka CNFs. We collaborate with the cncf/cnf-conformance test suite who work on the mechanics of the conformance tests.

> The goal for the group is to create a software conformance program that any application or network function implementation can use to demonstrate that they are conformant and interoperable with cloud native principles.

Reviewer: I think there are some areas where we would need to have some agreement "on-top" of cloud native best practices, like how to name Kubernetes resources or what kind of API we use to manage CNF-s on top that it should be a CRD (what is not more specific than "it should be an API") and similar things. Shouldn't we add these to the scope of the CNF WG or are these in the scope of the TUG? What others think?

Author: The request was to start with a narrower scope and only added when required. I expect that areas that are touched will at least be in the discussion for context even if they are not in scope for defining requirements.

## Mission Statement
Cloud Native Network Functions Working Group’s mission is to increase interoperability and standardization of cloud native workloads. It is committed to the following (aspirational) design ideals:

> - Portable - Cloud native workloads run everywhere -- public cloud, private cloud, bare metal, laptop -- with consistent behavior so that they are portable throughout the ecosystem as well as between development and production environments.

Reviewer 2: ```suggestion
- Portable - Cloud native workloads run everywhere -- public cloud, private cloud, bare metal, laptop -- with consistent functional behavior so that they are portable throughout the ecosystem as well as between development and production environments.
```

- Meet users partway. Many applications today are not cloud native, but have been working in production for decades. The WG doesn’t just cater to purely greenfield cloud-native applications, nor does it meet all users where they are. It focuses on cloud-native applications, but provides some mechanisms to facilitate migration of monolithic and legacy applications.
- Flexible. The cloud native technology ecosystem can be consumed a la carte and (in most cases) it does not prevent you from using your own solutions in lieu of built-in systems.

> - Extensible. Cloud native workloads should integrate into your environment and add the additional capabilities you need.

Reviewer 3: As is typical in this domain - do HW accelerators fit in here? I mean, can my FPGA custom Firmware NIC be exposed in a sane, non-binding way, so that my app can leverage it if it's there but can still do low-performance functions if it's not?

Author: In scope for discussion, right now, I presume would be that the app should be able to declaratively ask for some mechanism for using the FPGA custom Firmware NIC. 

That would imply that someone, even if it's not in this WG, should be figuring out how to expose access to those resources for that cloud native application to utilize.

Reviewer 4: It depends, infrastructure might provide the API to have access to the FPGA but a CNF would need to add the appropriate IP or runtime code associated with itself. Exemple node has an FPGA attached, it cannot load the FW microcode unless it knows there will be a FW CNF instantiated on that perticular node

> - Automatable. Cloud native workloads should aim to dramatically reduce the burden of manual operations. They support both declarative control by specifying users’ desired intent via an API, as well as imperative control to support higher-level orchestration and automation. The declarative approach is key to the ecosystem’s self-healing and autonomic capabilities.

Reviewer 2: ```suggestion
- Automatable. Cloud native workloads should aim to help dramatically reduce the burden of manual operations. They support both declarative control by specifying users’ desired intent via an API, as well as imperative control to support higher-level orchestration and automation. The declarative approach is key to the ecosystem’s self-healing and autonomic capabilities.
```

- Advance the state of the art. While the WG intends to drive the modernization of non-cloud-native applications, it also aspires to advance the cloud native and DevOps state of the art, such as in the participation of applications in their own management. Workloads should not be bound by the lowest common denominator of systems upon which they depend, such as container runtimes and cloud providers.

## In Scope

> - Definition of Cloud native Network Function (CNF)

Reviewer: Wasn't this already defined by the CNCF TUG?

Author: It will be reviewed and maintained in the CNF WG as a basis for the conformance definitions TBD by the broader group including leaders from other CNCF+K8s SIGs and WGs.

Reviewer 5: Yes, we are seeking to formalize the structure and make it more of a living document so as the state of the art in cloud native progresses so does the definition

- Cloud native conformance test requirements for CNFs 

> - Process around certifying CNF conformance

Reviewer 2: ```suggestion
- Process around certifying CNFs.
```

Author: Defining what a CNF is in scope. That is possibly different from certification. For the certifying I think we are trying to say how something adheres to cloud native best practices + principles.  

Do we want something more focused, such as:

> Process around certifying CNFs adherence to cloud native principles

if not then I like the suggestion from @rabi-abdel which is more general.

@caniszczyk @xmulligan

Author: @rabi-abdel how about 

> Process around certifying the cloud native conformance of CNFs

which is the intent of the original wording. Maybe the intent is given based on the program being about cloud native?

> - Feedback into other related groups and specifications to improve CNF use cases (e.g. SIG App Delivery, SIG Networking, CNI)

Reviewer 2: ```suggestion
- Feedback into other related groups and specification bodies to improve CNF use cases (e.g. SIG App Delivery, SIG Networking, CNI)
```

> - Feedback into other related groups and specification bodies to improve CNF use cases (e.g. SIG App Delivery, SIG Networking, CNI)

Reviewer: How about downstream communities, like CNTT?

Author: Sure. ORAN, TIP, and others as well.

Note the e.g. indicating just a few mentioned, not all.

- Publishing metrics/white papers
- Best Practices and General Recommendations

## Potential Future Scope

> - Cloud native conformance test requirements for Telco platforms (which run CNFs)

Reviewer 2: ```suggestion
- Cloud native conformance requirements for Telco platforms (which run CNFs)
```

> - Cloud native conformance test requirements for Telcom platforms (which run CNFs)

Reviewer 4: Where does the Telcom acronym comes from ? 1st time being used ... should be "Telco platforms"

> - Cloud native conformance test requirements for Telcom platforms (which run CNFs)

Reviewer 5: ```suggestion
- Cloud native conformance test requirements for Telco platforms (which run CNFs)
```

> - Cloud native conformance requirements for Telco platforms (which run CNFs)

Reviewer: This is in the scope of CNTT RA2, therefore we should not have it here.

Author: ATM it is out of scope for the WG. In the future it may decide to extend the scope, if a new CNCF group is not formed, which focuses on the layers under the telco applications. This would be focused on cloud native best practices/principles for infra/platform layers -- which is in scope for CNCF.

## Out of Scope

- Writing conformance tests or a test suite
- Building Tooling
- Promotion of specific tools
- Solving external dependencies

## Overlap and Relations with other Groups and Projects
The CNF WG sees itself as providing the upstream definition of what makes a telco application cloud native allowing downstream projects to create precise programs and/or implementations for their specific needs. Some of the groups who may utilize the CNF Conformance Programs deliverables are:

> - CNTT R2 - is focused on Kubernetes-based platforms and basic interoperability between platform and workloads. CNTT R2 has not determined if workload cloud native requirements are in scope for CNTT R2. It is expecting CNCF to provide testing for the cloud native requirements it has defined. 

Reviewer: I think CNF requirements related to the infrastructure are in scope for CNTT RA2. @tomkivlin what do you think?

Reviewer 6: @CsatariGergely I agree, but I think that's what "basic interoperability between platform and workloads" covers?

Author: CNTT RA2 is creating an opinionated K8s telco distro.

If it wants to have workload requirements beyond basic interoperability it can choose to include the CNF Conformance requirements in addition to what ever other requirements it wishes to add.

Reviewer: @tomkivlin yes, that is true, I'm more worried about the "CNTT R2 has not determined if workload cloud native requirements are in scope for CNTT R2" part. We have the (still empty) [Chapter 4.9](https://cntt.readthedocs.io/en/latest/ref_arch/kubernetes/chapters/chapter04.html#4.9) in RA2.

Reviewer 4: CNF requirements should remain outside of CNTT RA2 scope although it can relate and promote them. CNF conformance and requirements need to be also applicable to non CNTT-RA2 compliant platforms.

- OVP 2.0 (Cloud Native) - is interested in leveraging an upstream source for cloud native requirements and test results (like deliverables from the CNCF CNF WG) to be used in the OVP 2.0 Badging Program.

Telco applications and the workloads that are created with them are related to many topics in Cloud Native computing; therefore this WG may collaborate with many of the other CNCF and K8s SIGs, WGs, and projects. However, the following groups might have the largest potential interactions:

- CNCF SIG App Delivery
- CNCF SIG Security 
- CNCF SIG Network
- Kubernetes SIG Apps
- Kubernetes SIG Testing
- K8s Conformance WG

> 

Reviewer 4: what about k8s-sig-networking, k8s-sig-multitenancy, cncf sig service mesh, k8snetworkplumbing ?

Author: @dabernie, sure. Maybe it should be worded to give a few examples vs pointing out any that may be worked with more than others. There is a lot of overlap.

Author: @dabernie updated wording.

> A few of the groups with potential interactions/collaboration are:

> 

Reviewer 4: Does it only target these k8s-sigs ? should it not include some like multi-tenancy, networking ?

Author: @dabernie this is just an example set of the SIGs and groups we expect to collaborate with not a full list.

- Promotion of specific tools
- Solving external dependencies

## Overlap and Relations with other Groups and Projects
The CNF WG sees itself as providing the upstream definition of what makes a telco application cloud native allowing downstream projects to create precise programs and/or implementations for their specific needs. Some of the groups who may utilize the CNF Conformance Programs deliverables are:

- CNTT R2 - is focused on Kubernetes-based platforms and basic interoperability between platform and workloads. CNTT R2 has not determined if workload cloud native requirements are in scope for CNTT R2. It is expecting CNCF to provide testing for the cloud native requirements it has defined. 
- OVP 2.0 (Cloud Native) - is interested in leveraging an upstream source for cloud native requirements and test results (like deliverables from the CNCF CNF WG) to be used in the OVP 2.0 Badging Program.

Telco applications and the workloads that are created with them are related to many topics in Cloud Native computing; therefore this WG may collaborate with many of the other CNCF and K8s SIGs, WGs, and projects. A few of the groups with potential interactions/collaboration are:

> 

Reviewer 7: What about interaction with non-CNCF projects, E.g. ONAP? There might be a benefit in some sort of alignment (or at least mutual awareness) with ONAP. BTW- I volunteer to serve as aa liaison ;-)

Author: @rannyh, I expect there will be interactions with many non-CNCF orgs and projects. Ideally ONAP folks will participate in the CNF Working Group to provide input. 

Beyond that there are several ways to collaborate with testing itself including ONAP running tests from the test suite as part of onboarding CNFs as well as testing ONAP components themselves.

- CNCF SIG App Delivery
- CNCF SIG Security 
- CNCF SIG Network
- Kubernetes SIG Apps
- Kubernetes SIG Testing
- K8s Conformance WG

> 

Reviewer 7: What about interaction with non-CNCF projects, E.g. ONAP? There might be a benefit in some sort of alignment (or at least mutual awareness) with ONAP. BTW- I volunteer to serve as aa liaison ;-)
